src/PluginsSystem/Tools/Microsoft.Performance.Toolkit.Plugins.Cli/Program.cs

```diff
@@ -0,0 +1,226 @@
+using Microsoft.Performance.Toolkit.Plugins.Cli.Exceptions;
+using Microsoft.Performance.Toolkit.Plugins.Cli.Manifest;
+using Microsoft.Performance.Toolkit.Plugins.Cli.Metadata;
+using Microsoft.Performance.Toolkit.Plugins.Cli.Options;
+using Microsoft.Performance.Toolkit.Plugins.Cli.Packaging;
+using Microsoft.Performance.Toolkit.Plugins.Cli.Validation;
+using Microsoft.Performance.Toolkit.Plugins.Core.Metadata;
+using Microsoft.Performance.Toolkit.Plugins.Core.Serialization;
+using Microsoft.Performance.Toolkit.Plugins.Runtime.Package;
+
+namespace Microsoft.Performance.Toolkit.Plugins.Cli
+{
+    /// <summary>
+    ///     Represents the entry point for the CLI.
+    /// </summary>
+    public sealed class Program
+    {
+        /// <summary>
+        ///     Main entry point.
+        /// </summary>
+        /// <param name="args">
+        ///     The command line arguments.
+        /// </param>
+        /// <returns>
+        ///     A task whose result is the exit code. 0 on success; otherwise, non-zero.
+        /// </returns>
+        public static async Task<int> Main(string[] args)
+        {
+            ParserResult<object> result = Parser.Default.ParseArguments<PackOptions, MetadataGenOptions>(args);
+
+            ServiceProvider serviceProvider = new ServiceCollection()
+                .AddLogging(x => x.AddConsole())
+                .AddSingleton<IMetadataGenerator, MetadataGenerator>()
+                .AddSingleton<ISourceFilesProcessor, PluginSourceFilesProcessor>()
+                .AddSingleton<IPackageBuilder, ZipPluginPackageBuilder>()
+                .AddSingleton<ISerializer<PluginManifest>>(SerializationUtils.GetJsonSerializer<PluginManifest>(new JsonSerializerOptions
+                {
+                    WriteIndented = true,
+                    PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
+                }))
+                .AddSingleton<ISerializer<PluginMetadata>>(SerializationUtils.GetJsonSerializerWithDefaultOptions<PluginMetadata>())
+                .AddSingleton<ISerializer<PluginContentsMetadata>>(SerializationUtils.GetJsonSerializerWithDefaultOptions<PluginContentsMetadata>())
+                .AddSingleton<IManifestFileReader, ManifestReader>()
+                .AddSingleton<IManifestFileValidator, ManifestJsonSchemaValidator>()
+                .AddSingleton<IJsonSchemaLoader, LocalManifestSchemaLoader>()
+                .BuildServiceProvider();
+
+            ILogger logger = serviceProvider.GetRequiredService<ILogger<Program>>();
+
+            try
+            {
+                return result.MapResult(
+                    (PackOptions opts) => RunPack(opts, serviceProvider),
+                    (MetadataGenOptions opts) => RunMetadataGen(opts, serviceProvider),
+                    errs => HandleParseError(errs, serviceProvider));
+            }
+            catch (ArgumentValidationException ex)
+            {
+                logger.LogError("Invalid arguments. {0}", ex.Message);
```

Comment:
To me, having all these exceptions is a code smell. Are these truly "exceptional?" For instance, invalid arguments aren't an exception... we should expect and handle that case gracefully, returning early with a useful error message/code. Same with other invalid inputs (manifest, plugin content, etc.).

The issue I have with this pattern is the program's entry point is essentially an uber-class, needing to know about everything that could possibly go wrong while the program is executing. As more things are checked, this list grows.

I personally would get rid of these exceptions and just use facades to abstract out individual steps